src/topotoolbox/stream_object.py

```diff
@@ -1215,46 +1219,48 @@ def crslin(self, dem, k = 1, mingradient=0.0, attachheads=False, attachtomin=Fal
             ids = np.where(channelheads)
 
             # matrix with the channelheads on main diagonal
-            i_eq = sp.coo_matrix((channelheads[channelheads].astype(float),
+            i_eq = sp.csc_array((channelheads[channelheads].astype(float),
                                   (np.arange(nc), ids[0])), shape=(nc, nr))
-            i_eq = spmatrix(i_eq.data.tolist(), i_eq.row.tolist(
-            ), i_eq.col.tolist(), (nc, nr))  # convert to kvxopt matrix
-
-            z_eq = z[channelheads]  # store elevation values at channelheads
-            z_eq = matrix(z_eq)  # convert into kvxopt matrix
+            z_eq = z[channelheads]
 
         else:
-            i_eq = spmatrix([], [], [], (0, nr))
-            z_eq = matrix(0.0, (0,1))
+            i_eq = sp.csc_array((0, nr))
+            z_eq = np.zeros(0)
 
         # Inequality constraints
         # Gradient constraint
-        dd = np.array(1/(d[self.source]-d[self.target]))  # cellsize
+        delta = np.array(1/(d[self.source]-d[self.target]))  # cellsize
 
-        gradient = (sp.coo_matrix((dd, (self.source, self.target)), shape=(
-            nr, nr)) - sp.coo_matrix((dd, (self.source, self.source)), shape=(nr, nr))).tocoo()
-        gradient = spmatrix(gradient.data.tolist(), gradient.row.tolist(
-        ), gradient.col.tolist(), (nr, nr))  # convert to ckxopt matrix
+        gradient = sp.csc_array(
+            (delta, (np.arange(ne), self.target)),shape=(ne, nr)) - sp.csc_array(
+                (delta, (np.arange(ne), self.source)), shape=(ne, nr))
 
-        g_min = np.zeros((nr, 1))  # minimum gradient
+        g_min = np.zeros(nr)  # minimum gradient
         g_min[self.source] = mingradient
```

Comment:
The remaining errors are because `g_min` needs to have length `ne`, but you have left it the older way with `nr` elements. Change this line to:

```python
g_min = np.full(ne, mingradient)
```